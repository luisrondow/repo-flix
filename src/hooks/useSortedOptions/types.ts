import { TECHS, SORT_OPTIONS } from '../../utils/constants'

export type SortedOptions = {
  tech: (typeof TECHS)[number]
  sort: (typeof SORT_OPTIONS)[number]['value']
}[]
