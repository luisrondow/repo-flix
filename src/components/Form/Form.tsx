import { useForm } from 'react-hook-form'

import Input from '../Input'
import Button from '../Button'

import { FormProps } from './Form.types'
import { FORM_ERRORS } from '../../utils/constants'

const Form = <T extends object>(props: FormProps<T>) => {
  const { formOptions, onSubmit, isSubmitting, submitText } = props

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<T>()

  return (
    <div className="mt-12">
      <form className="flex flex-col" onSubmit={handleSubmit(onSubmit)}>
        {formOptions.map((option) => (
          <Input
            key={option.key}
            defaultValue={option.defaultValue}
            className="mb-8"
            label={option.label}
            placeholder={option.placeholder}
            type={option.type || 'text'}
            error={
              errors[option.key]?.type
                ? FORM_ERRORS.find((el) => el.value === errors[option.key]?.type)?.label
                : ''
            }
            {...register(option.key, {
              required: option.required,
              ...(option.pattern && { pattern: option.pattern }),
            })}
          />
        ))}
      </form>
      <Button
        type="submit"
        className="w-full"
        onClick={handleSubmit(onSubmit)}
        disabled={isSubmitting}
      >
        {isSubmitting ? 'Loading...' : submitText}
      </Button>
    </div>
  )
}

export default Form
