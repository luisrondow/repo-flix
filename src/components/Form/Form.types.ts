import { Path } from 'react-hook-form'

export interface FormOption<T> {
  key: Path<T>
  label: string
  placeholder: string
  defaultValue?: string
  required?: boolean
  type?: string
  pattern?: RegExp
  minLength?: number
}

export interface FormProps<T> {
  formOptions: FormOption<T>[]
  onSubmit: (values: T) => void
  isSubmitting: boolean
  submitText: string
}
