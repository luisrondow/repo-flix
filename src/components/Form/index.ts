export { default } from './Form'
export type { FormOption } from './Form.types'
