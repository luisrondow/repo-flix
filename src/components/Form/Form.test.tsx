import { render } from '@testing-library/react'
import Form from './Form'
import { FormOption } from './Form.types'
import { act } from 'react-dom/test-utils'

const mockFormOptions: FormOption<{
  email: string
  password: string
}>[] = [
  {
    key: 'email',
    label: 'Email',
    placeholder: 'Email',
    required: true,
    pattern: /^\S+@\S+$/i,
  },
  {
    key: 'password',
    label: 'Password',
    placeholder: 'Password',
    required: true,
    type: 'password',
  },
]
const mockOnSubmit = jest.fn()

describe('Form', () => {
  it('should render correctly', () => {
    const { container, getByLabelText, getByText } = render(
      <Form
        formOptions={mockFormOptions}
        onSubmit={mockOnSubmit}
        isSubmitting={false}
        submitText="Submit"
      />,
    )

    expect(container).toMatchSnapshot()

    mockFormOptions.forEach((option) => {
      expect(getByLabelText(option.label)).toBeInTheDocument()
    })

    expect(getByText('Submit')).toBeInTheDocument()
  })

  it('should render correctly when isSubmitting is true', () => {
    const { container, getByText } = render(
      <Form
        formOptions={mockFormOptions}
        onSubmit={mockOnSubmit}
        isSubmitting
        submitText="Submit"
      />,
    )

    expect(getByText('Loading...')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  // eslint-disable-next-line jest/no-disabled-tests
  it.skip('should fire the onSubit when submit form', () => {
    const { getByText } = render(
      <Form
        formOptions={mockFormOptions}
        onSubmit={mockOnSubmit}
        isSubmitting={false}
        submitText="Submit"
      />,
    )

    act(() => {
      getByText('Submit').click()
    })

    expect(mockOnSubmit).toHaveBeenCalled()
  })
})
