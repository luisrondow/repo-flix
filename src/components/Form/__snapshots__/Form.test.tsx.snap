// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render correctly 1`] = `
<div>
  <div
    class="mt-12"
  >
    <form
      class="flex flex-col"
    >
      <div
        class="mb-8"
      >
        <label>
          <h1
            class="mb-2 text-base text-white text-[#e0def4] font-roboto text-2xl font-bold sm:text-4xl"
          >
            Email
          </h1>
          <input
            class="h-12 w-full rounded-md bg-base px-4 text-white"
            name="email"
            placeholder="Email"
            type="text"
            value=""
          />
          <p
            class="mt-2 text-sm font-bold text-red text-[#e0def4] font-roboto"
          />
        </label>
      </div>
      <div
        class="mb-8"
      >
        <label>
          <h1
            class="mb-2 text-base text-white text-[#e0def4] font-roboto text-2xl font-bold sm:text-4xl"
          >
            Password
          </h1>
          <input
            class="h-12 w-full rounded-md bg-base px-4 text-white"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <p
            class="mt-2 text-sm font-bold text-red text-[#e0def4] font-roboto"
          />
        </label>
      </div>
    </form>
    <button
      class="mt-8 h-12 rounded-md bg-white font-semibold transition-all duration-200 hover:bg-gray-light w-full"
      type="submit"
    >
      Submit
    </button>
  </div>
</div>
`;

exports[`Form should render correctly when isSubmitting is true 1`] = `
<div>
  <div
    class="mt-12"
  >
    <form
      class="flex flex-col"
    >
      <div
        class="mb-8"
      >
        <label>
          <h1
            class="mb-2 text-base text-white text-[#e0def4] font-roboto text-2xl font-bold sm:text-4xl"
          >
            Email
          </h1>
          <input
            class="h-12 w-full rounded-md bg-base px-4 text-white"
            name="email"
            placeholder="Email"
            type="text"
            value=""
          />
          <p
            class="mt-2 text-sm font-bold text-red text-[#e0def4] font-roboto"
          />
        </label>
      </div>
      <div
        class="mb-8"
      >
        <label>
          <h1
            class="mb-2 text-base text-white text-[#e0def4] font-roboto text-2xl font-bold sm:text-4xl"
          >
            Password
          </h1>
          <input
            class="h-12 w-full rounded-md bg-base px-4 text-white"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <p
            class="mt-2 text-sm font-bold text-red text-[#e0def4] font-roboto"
          />
        </label>
      </div>
    </form>
    <button
      class="mt-8 h-12 rounded-md bg-white font-semibold transition-all duration-200 hover:bg-gray-light w-full"
      disabled=""
      type="submit"
    >
      Loading...
    </button>
  </div>
</div>
`;
